app/controllers/course/learning_map_controller.rb

```diff
@@ -1,15 +1,219 @@
+        node_ids_to_parents[node_id].push(parent)
+      end
+    end
+
+    { node_ids_to_children: node_ids_to_children, node_ids_to_parents: node_ids_to_parents }
+  end
+
+  def init_all_node_relations
+    { node_ids_to_children: @conditionals.map { |conditional| [get_node_id(conditional), []] }.to_h,
+      node_ids_to_parents: @conditionals.map { |conditional| [get_node_id(conditional), []] }.to_h }
+  end
+
+  def map_condition_to_parent(condition)
+    type = condition.actable_type.demodulize
+    typed_condition = Object.const_get("Course::Condition::#{type}").preload(:actable).find(condition.actable_id)
+    id = "#{type.downcase}-#{typed_condition.send("#{type.downcase}_id")}"
+
+    { id: id, is_satisfied: typed_condition.satisfied_by?(current_course_user) }
+  end
+
+  def generate_nodes_from_conditionals(all_node_relations)
+    node_ids_to_children = all_node_relations[:node_ids_to_children]
+    node_ids_to_parents = all_node_relations[:node_ids_to_parents]
+
+    @conditionals.map do |conditional|
+      id = get_node_id(conditional)
+
+      conditional.attributes.merge({
+        id: id, unlocked: conditional.conditions_satisfied_by?(current_course_user),
+        children: node_ids_to_children[id], satisfiability_type: conditional.satisfiability_type,
+        course_material_type: conditional.class.name.demodulize.downcase,
+        content_url: url_for([current_course, conditional]), parents: node_ids_to_parents[id]
+      }).symbolize_keys
+    end
+  end
+
+  def generate_node_depths(nodes)
+    toposorted_nodes = toposort(nodes)
+    depths = init_depths(nodes)
+
+    toposorted_nodes.each do |node|
+      node_id = node[:id]
+
+      node[:children].each do |child|
+        child_id = child[:id]
+        depths[child_id] = depths[node_id] + 1 if depths[child_id] < depths[node_id] + 1
+      end
+    end
+
+    nodes.map { |node| node.merge({ depth: depths[node[:id]] }) }
+  end
+
+  def init_depths(nodes)
+    nodes.map { |node| [node[:id], node[:parents].empty? ? 0 : NEGATIVE_INF] }.to_h
+  end
+
+  def toposort(nodes)
+    visited_node_ids = Set.new
+    post_order_nodes = []
+    node_ids_to_nodes = nodes.map { |node| [node[:id], node] }.to_h
+
+    nodes.each do |node|
+      dfs(node, node_ids_to_nodes, visited_node_ids, post_order_nodes) unless visited_node_ids.include?(node[:id])
+    end
+
+    post_order_nodes.reverse
+  end
+
+  def dfs(node, node_ids_to_nodes, visited_node_ids, post_order_nodes)
+    visited_node_ids.add(node[:id])
+
+    node[:children].each do |child|
+      dfs(node_ids_to_nodes[child[:id]], node_ids_to_nodes, visited_node_ids, post_order_nodes) unless
+        visited_node_ids.include?(child[:id])
+    end
+
+    post_order_nodes.push(node)
+  end
+
+  def parent_and_node_id_pair_params
+    params.permit(:parent_node_id, :node_id)
+  end
+
+  def node_params
+    params.permit(:node_id)
+  end
+
+  def get_node_id(conditional)
```

Comment:
I think this part onwards is quite interesting haha, but feels a little bit on the hacky side as well. Might be less hacky to use some form of enum / map from a constant to a class (which, tbh, is similar to what you're doing here), then the frontend can send the constant over.

I guess the reason for this is more of security? Allowing users to directly determine what class you're constructing with (albeit limited to the namespace `Course::Condition`) feels a bit dangerous to me. Validation should be done accordingly.